lgc/builder/BuilderImpl.cpp

```diff
@@ -323,162 +323,239 @@ BranchInst *BuilderImplBase::createIf(Value *condition, bool wantElse, const Twi
-Instruction *BuilderImplBase::createWaterfallLoop(Instruction *nonUniformInst, ArrayRef<unsigned> operandIdxs,
-                                                  const Twine &instName) {
-#if !defined(LLVM_HAVE_BRANCH_AMD_GFX)
-#warning[!amd-gfx] Waterfall feature disabled
-  errs() << "Generating invalid waterfall loop code\n";
-  return nonUniformInst;
-#else
-  assert(operandIdxs.empty() == false);
-
-  // For each non-uniform input, try and trace back through a descriptor load to
-  // find the non-uniform index used in it. If that fails, we just use the
-  // operand value as the index.
-  //
-  // Note: this code has to cope with relocs as well, this is why we have to
-  // have a worklist of instructions to trace back
-  // through. Something like this:
-  // %1 = call .... @lgc.descriptor.set(...)          ;; Known uniform base
-  // %2 = call .... @llvm.amdgcn.reloc.constant(...)  ;; Known uniform reloc constant
-  // %3 = ptrtoint ... %1 to i64
-  // %4 = zext ... %2 to i64
-  // %5 = add i64 %3, %4
-  // %6 = inttoptr i64 %5 to ....
-  // %7 = bitcast .... %6 to ....
-  // %8 = getelementptr .... %7, i64 %offset
-  //
-  // As long as the base pointer %7 can be traced back to a descriptor set and
-  // reloc we can infer that it is truly uniform and use the gep index as the waterfall index safely
-  //
-
-  SmallVector<Value *, 2> nonUniformIndices;
-  SmallVector<Value *, 2> worklist;
-
-  for (unsigned operandIdx : operandIdxs) {
-    Value *nonUniformVal = nonUniformInst->getOperand(operandIdx);
-    if (auto load = dyn_cast<LoadInst>(nonUniformVal)) {
-      Value *base = load->getOperand(0);
-      Value *index = nullptr;
-      for (;;) {
-        if (auto bitcast = dyn_cast<BitCastInst>(base)) {
-          base = bitcast->getOperand(0);
-          continue;
-        }
-        if (auto gep = dyn_cast<GetElementPtrInst>(base)) {
-          if (gep->hasAllConstantIndices()) {
-            base = gep->getPointerOperand();
-            continue;
-          }
-          // Variable GEP, to provide the index for the waterfall.
-          if (index || gep->getNumIndices() != 1)
-            break;
-          index = *gep->idx_begin();
+// Note: this code has to cope with relocs as well, this is why we have to
+// have a worklist of instructions to trace back
+// through. Something like this:
+// %1 = call .... @lgc.descriptor.set(...)          ;; Known uniform base
+// %2 = call .... @llvm.amdgcn.reloc.constant(...)  ;; Known uniform reloc constant
+// %3 = ptrtoint ... %1 to i64
+// %4 = zext ... %2 to i64
+// %5 = add i64 %3, %4
+// %6 = inttoptr i64 %5 to ....
+// %7 = bitcast .... %6 to ....
+// %8 = getelementptr .... %7, i64 %offset
+//
+// As long as the base pointer %7 can be traced back to a descriptor set and
+// reloc we can infer that it is truly uniform and use the gep index as the waterfall index safely
+//
+static Value *traceNonUniformIndex(Value *nonUniformVal) {
+  if (auto load = dyn_cast<LoadInst>(nonUniformVal)) {
```

Comment:
Can we return early so that we reduce code nesting?

I mean something like:
```
auto load = dyn_cast<LoadInst>(nonUniformVal);
if (!load)
  return nonUniformVal;
```

See https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code and https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html

Reply:
Done